src/sentry/replays/endpoints/project_replay_recording_segment_index.py

```diff
@@ -18,15 +26,80 @@ def get(self, request: Request, project, replay_id: str) -> Response:
         ):
             return self.respond(status=404)
 
-        return self.paginate(
-            request=request,
-            queryset=ReplayRecordingSegment.objects.filter(
-                project_id=project.id,
-                replay_id=replay_id.replace("-", ""),
-            ),
-            order_by="segment_id",
-            on_results=lambda x: {
-                "data": serialize(x, request.user, ReplayRecordingSegmentSerializer())
-            },
-            paginator_cls=OffsetPaginator,
-        )
+        if request.GET.get("download") is not None:
+            return self.paginate(
+                request=request,
+                queryset=ReplayRecordingSegment.objects.filter(
+                    project_id=project.id,
+                    replay_id=replay_id.replace("-", ""),
+                ),
+                order_by="segment_id",
+                on_results=self.on_download_results,
+                response_cls=StreamingHttpResponse,
+                response_kwargs={"content_type": "application/json"},
+                paginator_cls=OffsetPaginator,
+            )
+
+        else:
+            return self.paginate(
+                request=request,
+                queryset=ReplayRecordingSegment.objects.filter(
+                    project_id=project.id,
+                    replay_id=replay_id.replace("-", ""),
+                ),
+                order_by="segment_id",
+                on_results=lambda x: {
+                    "data": serialize(x, request.user, ReplayRecordingSegmentSerializer())
+                },
+                paginator_cls=OffsetPaginator,
+            )
+
+    def on_download_results(self, results):
+        """
+        get the files associated with the segment range requested. prefetch the files
+        in a threadpool.
+        """
+        recording_segment_files = File.objects.filter(id__in=[r.file_id for r in results])
+        # TODO: deflate files as theyre fetched, instead of having
+        # to wait untill all of them are complete before starting work.
+        with ThreadPoolExecutor(max_workers=4) as exe:
+            file_objects = exe.map(
+                lambda file: file.getfile(prefetch=True), recording_segment_files
+            )
+
+        return iter(self.segment_generator(list(file_objects)))
+
+    def segment_generator(self, recording_segments):
+        """
+        streams a JSON object made of replay recording segments.
+        the segments are individual json objects, and we build a list around them.
+        they are also default compressed, so deflate them if needed.
+        """
+        yield "["
+
+        for i, file in enumerate(recording_segments):
+            if self.is_compressed(file):
+                yield from self.decompress_blob_stream(file)
+            else:
+                yield file.read().decode("utf-8")
+
+            if i < len(recording_segments) - 1:
+                yield ","
+
+        yield "]"
```

Comment:
Do we need to serialize this as JSON?

Reply:
if we don't do it here the client has to its own parsing / serialization, and so it here can clean up that client side logic.